### PR TITLE
ensure work with given replicas

### PR DESCRIPTION
**Pull request by qianjun1993**
Signed-off-by: junqian <junqian@tencent.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Set replicas in work according to the info in target clusters

**Which issue(s) this PR fixes**:
Part of #344

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

NONE

```release-note

```

### CODE REVIEW

File: pkg/util/helper/binding.go [hunk 1]
```diff
@@ -153,12 +163,20 @@ func FetchWorkload(dynamicClient dynamic.Interface, restMapper meta.RESTMapper,
...
-	referenceRSP, desireReplicaInfos, err := calculateReplicasIfNeeded(c, workload, clusterNames)
-	if err != nil {
-		klog.Errorf("Failed to get ReplicaSchedulingPolicy for %s/%s/%s, err is: %v", workload.GetKind(), workload.GetNamespace(), workload.GetName(), err)
-		return err
+func EnsureWork(c client.Client, workload *unstructured.Unstructured, clusterNames []string, clusters []workv1alpha1.TargetCluster,
+	overrideManager overridemanager.OverrideManager, binding metav1.Object, scope apiextensionsv1.ResourceScope) error {
+	var desireReplicaInfos map[string]int64
+	var referenceRSP *v1alpha1.ReplicaSchedulingPolicy
+	var err error
+	isBindingReplicasReady := IsBindingReplicasReady(clusters)
```
Comment (mrlihanbo) [inline pkg/util/helper/binding.go:171]: No need to check if resource without 'replicas' ?

Comment (qianjun1993) [reply pkg/util/helper/binding.go:171]: if resource is without replicas the replicas will be zero in clusters。 This judgement will return false .

File: pkg/util/helper/binding.go [hunk 2]
```diff
@@ -153,12 +163,20 @@ func FetchWorkload(dynamicClient dynamic.Interface, restMapper meta.RESTMapper,
 // EnsureWork ensure Work to be created or updated.
 //nolint:gocyclo
 // Note: ignore the cyclomatic complexity issue to get gocyclo on board. Tracked by: https://github.com/karmada-io/karmada/issues/460
-func EnsureWork(c client.Client, workload *unstructured.Unstructured, clusterNames []string, overrideManager overridemanager.OverrideManager,
-	binding metav1.Object, scope apiextensionsv1.ResourceScope) error {
-	referenceRSP, desireReplicaInfos, err := calculateReplicasIfNeeded(c, workload, clusterNames)
-	if err != nil {
-		klog.Errorf("Failed to get ReplicaSchedulingPolicy for %s/%s/%s, err is: %v", workload.GetKind(), workload.GetNamespace(), workload.GetName(), err)
-		return err
+func EnsureWork(c client.Client, workload *unstructured.Unstructured, clusterNames []string, clusters []workv1alpha1.TargetCluster,
```
Comment (RainbowMango) [inline pkg/util/helper/binding.go:166]:
```suggestion
func EnsureWork(c client.Client, workload *unstructured.Unstructured, clusterNames []string, scheduleResult []workv1alpha1.TargetCluster,
```
Give parameter a significant name.

File: pkg/util/helper/binding.go [hunk 3]
```diff
@@ -69,6 +69,16 @@ func IsBindingReady(targetClusters []workv1alpha1.TargetCluster) bool {
...
+// IsBindingReplicasReady will check if resourceBinding/clusterResourceBinding is ready to build Work with given replicas
+func IsBindingReplicasReady(targetClusters []workv1alpha1.TargetCluster) bool {
+	for _, targetCluster := range targetClusters {
+		if targetCluster.Replicas > 0 {
+			return true
+		}
+	}
+	return false
+}
+
```
Comment (RainbowMango) [inline pkg/util/helper/binding.go:81]:
```suggestion
// HasScheduledReplica checks if the scheduler has assigned replicas for each cluster.
func HasScheduledReplica(scheduleResult []workv1alpha1.TargetCluster) bool {
	for _, clusterResult := range scheduleResult {
		if clusterResult.Replicas > 0 {
			return true
		}
	}
	return false
}
```

File: pkg/util/helper/binding.go [hunk 4]
```diff
@@ -279,6 +297,14 @@ func EnsureWork(c client.Client, workload *unstructured.Unstructured, clusterNam
 	return nil
 }
 
+func changeTargetClusterToMap(targetCluster []workv1alpha1.TargetCluster) map[string]int64 {
```
Comment (RainbowMango) [inline pkg/util/helper/binding.go:300]:
```suggestion
func transScheduleResultToMap(scheduleResult []workv1alpha1.TargetCluster) map[string]int64 {
```

**Review by RainbowMango — CHANGES_REQUESTED** [review]

File: pkg/util/helper/binding.go (hunk 2 above)
Comment (qianjun1993) [reply pkg/util/helper/binding.go:166]: ok

File: pkg/util/helper/binding.go (hunk 3 above)
Comment (qianjun1993) [reply pkg/util/helper/binding.go:81]: ok

File: pkg/util/helper/binding.go (hunk 4 above)
Comment (qianjun1993) [reply pkg/util/helper/binding.go:300]: ok

**Review by RainbowMango — APPROVED** [review]

**Pull request merged**